### DISCUSSION TITLE
Large error in shape derivatives using checkTotals

### DISCUSSION THREAD

**Discussion by EdoAlvarezR** [mdolab/dafoam#515] [Bug]
Hi all,

Running the checkTotals task in the `NACA0012_Airfoil/incompressible` example points out that the shape derivatives wrt CL and CD have an error of ~100% with adjoint compared to finite difference. Is this expected?

I tried varying the finite difference step (from 1e-3 down to 1e-6 and up to 1e-1) and running it single-threaded, but the results were the same.

Here is the output of checkTotals and I'm attaching the case files (I am running everything in the Docker container from [here](https://dafoam.github.io/mydoc_get_started_download_docker.html)):

```
-----------------
Total Derivatives
-----------------

  Full Model: 'cruise.aero_post.CD' wrt 'aoa'
    Analytic Magnitude: 2.290762e-03
          Fd Magnitude: 2.291458e-03 (fd:central)
    Absolute Error (Jan - Jfd) : 6.964599e-07

    Relative Error (Jan - Jfd) / Jfd : 3.039374e-04 *

    Raw Analytic Derivative (Jfor)
[[0.0022907620224127]]

    Raw FD Derivative (Jfd)
[[0.0022914584822953]]
 - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -
  Full Model: 'cruise.aero_post.CD' wrt 'shape'
    Analytic Magnitude: 1.222483e-01
          Fd Magnitude: 2.209352e-01 (fd:central)
    Absolute Error (Jan - Jfd) : 1.840337e-01 *

    Relative Error (Jan - Jfd) / Jfd : 8.329759e-01 *

    Raw Analytic Derivative (Jfor)
[[ 0.0328568883043689  0.0326639317436422  0.0312709134803219
   0.0310666972689299 -0.0056783077221979 -0.005687546852263
  -0.0106064513358596 -0.01063232872407    0.0030197828103813
   0.0030310845338644  0.0058844708185029  0.0058726396359112
   0.0182145968549164  0.0182116476056585  0.0201410566277283
   0.0201261089439979 -0.0467618617978982 -0.0468257436373487
  -0.0480526803045212 -0.0481148982540672]]

    Raw FD Derivative (Jfd)
[[-0.0091725522642818  0.069489707359093  -0.0214847865056207
   0.0851767855545198  0.011439412579028  -0.0216047842836318
   0.002467054763402  -0.021386054144358   0.0431773095445962
  -0.0366597778229174  0.0585011972288445 -0.0462236983298929
   0.0684294071055049 -0.0309587252531998  0.0842275379685464
  -0.0429308786014602 -0.0235400972487074 -0.071371781989809
  -0.0226131434607026 -0.0749440683843794]]
 - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -
  Full Model: 'cruise.aero_post.CL' wrt 'aoa'
    Analytic Magnitude: 9.017470e-02
          Fd Magnitude: 9.016151e-02 (fd:central)
    Absolute Error (Jan - Jfd) : 1.318475e-05 *

    Relative Error (Jan - Jfd) / Jfd : 1.462347e-04 *

    Raw Analytic Derivative (Jfor)
[[0.0901746992904221]]

    Raw FD Derivative (Jfd)
[[0.0901615145449171]]
 - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -
  Full Model: 'cruise.aero_post.CL' wrt 'shape'
    Analytic Magnitude: 6.689883e+00
          Fd Magnitude: 7.103547e+00 (fd:central)
    Absolute Error (Jan - Jfd) : 2.278859e+00 *

    Relative Error (Jan - Jfd) / Jfd : 3.208058e-01 *

    Raw Analytic Derivative (Jfor)
[[ 0.0218795148654059  0.0203794604005753  0.2382735846068095
   0.2355469088986002  0.5183449206272644  0.5175561124222596
   0.6942199406144685  0.6928504752727487  0.6099535577914995
   0.6096388811918302  0.610189053259745   0.6094565329159577
   1.5856994216685898  1.5852603910663736  1.423928317099728
   1.4229900258957264 -2.6849018315728843 -2.6856768860462785
  -3.01237130728272   -3.0132170736956496]]

    Raw FD Derivative (Jfd)
[[ 0.4961712375509535 -0.4067243597126549  0.6965088416700098
  -0.2492606217513185  1.1862741351057196 -0.1541205621186919
   1.4019955851673274 -0.0240488442540538  1.0090280137495142
   0.2162879467605592  1.0215203956367702  0.205699799049313
   1.525942193354581   1.6690127780087494  1.3551614148382782
   1.5139013290564947 -3.3602126951474816 -2.0404804905111007
  -3.6959421949997306 -2.3657059884398564]]

 - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -
```

[naca0012-incompressible.zip](https://github.com/mdolab/dafoam/files/13326786/naca0012-incompressible.zip)

**Comment by friedenhe** [comment]
The error you saw is expected. This is caused by the finite-difference method for this special 2D case. When we use the finite-difference method and perturb one FFD point, it will create asymmetry geometry in the z direction and violate the symmetry boundary condition. So in this case, we should not trust the finite-difference totals. Please use the forward AD method to compute the ref derivatives for the shape variables. Check here: https://dafoam.github.io/mydoc_verifications_dasimplefoam.html. Anyway, I believe the adjoint is still accurate, but the finite-difference gradients for the shape variables are NOT.

**Reply by EdoAlvarezR** [reply]
Great! Thanks for pointing that out. 

I am seeing a similar error in the check when I set up an actuator disk (same 2D simulation) with the "direction" as a design variable (tilt angle of the rotor blowing on the airfoil). Do you think its also caused by the finite difference or perhaps the adjoint is not correctly connected internally in the derivative chain with the actuator disk? (I'll check it comparing to AD here shortly)

**Reply by EdoAlvarezR** [reply]
Also, great job developing DAFoam Ping! It's been pretty straight forward to pick it up and set up my own cases following the tutorials.

**Comment by friedenhe** [comment]
It is good to hear that DAFoam is useful. Regarding the inaccurate `direction` derivs, I don't think we can use `direction` as the design variable. We have 10 design variables for the actuator disk now, but `direction` is not one of them. Check here: https://github.com/mdolab/dafoam/blob/767274cdc4ca6c7de7649c88373c2c6611f1dea4/src/adjoint/DAFvSource/DAFvSource.C#L147 .

**Reply by EdoAlvarezR** [reply]
That explains why I had been struggling so much to get the optimizer to converge! Could you give me some pointer on how should I go about modifying the code if I wanted to add `direction` as a design variable?

Also, I started a new discussion with something that might be related: [LINK](https://github.com/mdolab/dafoam/discussions/524#discussion-5877165)

**Comment by friedenhe** [comment]
@EdoAlvarezR `direction` is added as a design variable in this commit: 234fbd32dbb378bbd061fd1d70cb032e564ebbd4. See an example of the new script: https://github.com/mdolab/dafoam/blob/234fbd32dbb378bbd061fd1d70cb032e564ebbd4/tests/runTests_DASimpleFoam.py#L172
